core/src/main/java/org/eclipse/hono/util/TenantObject.java

```diff
@@ -641,4 +578,172 @@ public OptionalInt getTraceSamplingPriority(final String authId) {
+                subjectNames.add(new X500Principal(subjectDn));
+            });
+            return subjectNames;
+        }
+    }
+
+    /**
+     * Get the list of configured trusted CAs for this tenant.
+     * 
+     * @return The list of configured trusted CAs or {@code null} if this tenant has no configured trusted CA.
+     *
+     */
+    @JsonIgnore
+    public List<JsonObject> getTrustedCAs() {
+
+        if (trustConfigurations == null) {
+            return null;
+        } else {
+            return trustConfigurations.values().stream().flatMap(Collection::stream).collect(Collectors.toList());
+        }
+
+    }
+
+    @JsonIgnore
+    private TrustAnchor getTrustAnchor(final JsonObject trustedCa) throws GeneralSecurityException {
+
+        final String subjectDn = getProperty(trustedCa, TenantConstants.FIELD_PAYLOAD_SUBJECT_DN, String.class);
+        final String encodedKey = getProperty(trustedCa, TenantConstants.FIELD_PAYLOAD_PUBLIC_KEY, String.class);
+        final String type = getProperty(trustedCa, TenantConstants.FIELD_PAYLOAD_KEY_ALGORITHM, String.class,
+                "RSA");
+
+        final Instant notBefore = getNotBefore(trustedCa);
+        final Instant notAfter = getNotAfter(trustedCa);
+        final Instant now = Instant.now();
+
+        if (subjectDn == null || encodedKey == null || notBefore == null || notAfter == null) {
+            return null;
+        } else if (now.isBefore(notBefore) || now.isAfter(notAfter)) {
+            // trusted CA is either too early for use or expired
+            return null;
+        } else {
+            final PublicKey pubKey = getPublicKey(encodedKey, type);
+            return new TrustAnchor(subjectDn, pubKey, null);
+        }
+    }
+
+    /**
+     * Add a trusted CA configuration for this tenant.
+     * 
+     * @param trustedCa The trust configuration to add.
+     * 
+     * @return This tenant so the API can be used fluently.
+     */
+    private TenantObject addTrustedCA(final JsonObject trustedCa) {
+
+        if (!isValidTenantConfig(trustedCa)) {
+            throw new IllegalArgumentException("invalid tenant configuration");
+        }
+
+        if (trustConfigurations == null) {
+            trustConfigurations = new HashMap<>();
+        }
+
+        final String subjectDn = getProperty(trustedCa, TenantConstants.FIELD_PAYLOAD_SUBJECT_DN, String.class);
+
+        final List<JsonObject> trustedCas = trustConfigurations.getOrDefault(subjectDn, new ArrayList<>());
+
+        trustedCas.add(trustedCa);
+        trustConfigurations.put(subjectDn, trustedCas);
+        return this;
+    }
+
+    private Instant getNotBefore(final JsonObject trustedCa) {
+        Instant instant = null;
+        final String strValue = getProperty(trustedCa, TenantConstants.FIELD_PAYLOAD_NOT_BEFORE, String.class);
+        if (strValue != null) {
+            instant = DateTimeFormatter.ISO_INSTANT.parse(strValue, Instant::from);
```

Comment:
IMHO this will fail to parse `2019-09-23T13:15:01+02:00` ...
I believe that you need to use `DateTimeFormatter.ISO_OFFSET_DATE_TIME` for that to work.

Maybe you can add a corresponding test case and see if it works.

Reply:
ok

Reply:
@sophokles73

> IMHO this will fail to parse 2019-09-23T13:15:01+02:00

but the device registry should not store this datetime value because it is not supported by Hono. As of  #1463, Hono now explicitly uses `DateTimeFormatter.ISO_INSTANT` for all its date time values (including the management API). the date time fields are now annotated with `@HonoTimestamp` and this annotation internally uses `DateTimeFormatter.ISO_INSTANT` as the supported date time format.

so I think the registry should reject storing such datetime values, WDYT?

Reply:
I do not think that's true. The Management API supports clients specifying *not-before* values like `2019-09-23T13:15:01+02:00`. These are correctly parsed (see `SecretsTest`) and then *normalized* to UTC. The `TenantObject` therefore should support this too FMPOV

Reply:
@sophokles73 Its true that clients can specify such values, but then all values are converted to `DateTimeFormatter.ISO_INSTANT` before they are transported to the backend service

 if you take a look at`SecretsTests#testDateFormats()`  you will see that the decoded `not-before` and `not-after` properties are not exactly in the same form as given by the client. 

> The TenantObject therefore should support this too FMPOV

ok, let me think about this :)

Reply:
> Its true that clients can specify such values, but then all values are converted to DateTimeFormatter.ISO_INSTANT before they are transported to the backend service.

What do mean with *before they are transported to the backend service*? The data is contained in a JSON object that is POSTed to the device registry's HTTP endpoint ...

Reply:
> if you take a look atSecretsTests#testDateFormats() you will see that the decoded not-before and not-after properties are not exactly in the same form as given by the client.

they don't need to. it is only important that the registry can parse these encoded instants. The same holds true for the TenantClient. It needs to be able to parse such encoded instants. It may then (internally) always us UTC to represent them, though ...

Reply:
I see your point! I think the way to approach this would be to define a specialise class in `org.eclipse.hono.util` to represent the domain object for a trusted certificate authority. Then I can annotate the `not-before` and `not-after` instances of the class using `@HonoTimestamp` so that Jackson can handle the parsing of the properties properly (as done in the management API)

Reply:
That would be an option, I guess. Based on our recent approach for sharing the low level value objects, you may even simply move the existing `TrustedCertificateAuthority` class from service-base to core and re-use/modify it accordingly. Remember that the Management API will also need to accept a list of CAs instead of just a single one ...

Reply:
yes, thats what am doing actually!

> Remember that the Management API will also need to accept a list of CAs instead of just a single one ...

yes, it already does that in this PR

Reply:
sure

Reply:
@sophokles73 I have been struggling with the following problem since in the morning and its not really funny :-)

for some reason, deserialising a `JsonObject` to a `org.eclipse.hono.util.TrustedCertificateAuthoritory` does not correctly parse a `datetime` with `offset` to an instant and this problem only happens in the `hono-core` module

example, giving this json:

```Java
final JsonObject json = new JsonObject()
                .put("subject-dn", "cn=test")
                .put("public-key", "notpublicKey".getBytes())
                .put("not-before", "2019-04-05T13:45:07-04:00")
                .put("not-after", "2037-06-01T14:00:00Z");
```
I want to convert it to an instance of `org.eclipse.hono.util.TrustedCertificateAuthoritory`,  so I did: `final TrustedCertificateAuthority ca = json.mapTo(TrustedCertificateAuthority.class);`

however, when I read the instant value using `ca.getNotBefore()` I get `2019-04-05T13:45:07Z`, the `time` value(i.e  13:45) is wrong.

The irony is when I run the same exact code inside of package `org.eclipse.hono.service.management.tenant` :

`final org.eclipse.hono.util.TrustedCertificateAuthority ca = json.mapTo(org.eclipse.hono.util.TrustedCertificateAuthority.class);` then the result I get is `2019-04-05T17:45:07Z` (which is the correct value, IMO).

I don't understand why deserialising a JSON to a `org.eclipse.hono.util.TrustedCertificateAuthority` should yield different results in different modules. I removed the following dependencies in the `core/pom.xml`, which I was suspecting was causing the problem in the `core` module, but with no success:

```xml
    <dependency>
      <groupId>com.fasterxml.jackson.core</groupId>
      <artifactId>jackson-databind</artifactId>
      <version>${jackson.version}</version>
    </dependency>
    <dependency>
      <groupId>com.fasterxml.jackson.core</groupId>
      <artifactId>jackson-annotations</artifactId>
      <version>${jackson.version}</version>
```

have you ever had this problem with the Jackson library?

BTW: the `TrustedCertificateAuthority` is exactly the same as in the `service-base` and `core` modules.

Reply:
Can you post the code of your `TrustedCertifcateAuthority` class in core?

Reply:
```java
package org.eclipse.hono.util;

import java.time.Instant;

import org.eclipse.hono.annotation.HonoTimestamp;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonInclude.Include;

/**
 * A trusted CA configuration.
 *
 */
@JsonInclude(value = Include.NON_NULL)
public class TrustedCertificateAuthority {

    @JsonProperty(value = TenantConstants.FIELD_PAYLOAD_SUBJECT_DN, required = true)
    private String subjectDn;

    @JsonProperty(TenantConstants.FIELD_PAYLOAD_PUBLIC_KEY)
    private byte[] publicKey;

    @JsonProperty(TenantConstants.FIELD_PAYLOAD_CERT)
    private byte[] certificate;

    @JsonProperty(TenantConstants.FIELD_PAYLOAD_NOT_BEFORE)
    @HonoTimestamp
    private Instant notBefore;

    @JsonProperty(TenantConstants.FIELD_PAYLOAD_NOT_AFTER)
    @HonoTimestamp
    private Instant notAfter;

    @JsonProperty(TenantConstants.FIELD_PAYLOAD_KEY_ALGORITHM)
    private String keyAlgorithm;

    /**
     * Sets the subject-dn configuration property.
     *
     * @param subjectDn The subject-dn property to set.
     * @return          a reference to this for fluent use.
     */
    public TrustedCertificateAuthority setSubjectDn(final String subjectDn) {
        this.subjectDn = subjectDn;
        return this;
    }

    public String getSubjectDn() {
        return subjectDn;
    }

    /**
     * Sets the public-key configuration property.
     *
     * @param publicKey The public-key property of the trusted root certificate.
     * @return          a reference to this for fluent use.
     */
    public TrustedCertificateAuthority setPublicKey(final byte[] publicKey) {
        this.publicKey = publicKey;
        return this;
    }

    public byte[] getPublicKey() {
        return publicKey;
    }

    /**
     * Sets the cert configuration property.
     *
     * @param certificate The cert property of the trusted root certificate.
     * @return            a reference to this for fluent use.
     */
    public TrustedCertificateAuthority setCertificate(final byte[] certificate) {
        this.certificate = certificate;
        return this;
    }

    public byte[] getCertificate() {
        return certificate;
    }

    /**
     * Sets the not-before property.
     * 
     * @param notBefore The not-before value to assign.
     * 
     * @return A reference to this API so it can be used fluently.
     */
    public TrustedCertificateAuthority setNotBefore(final Instant notBefore) {
        this.notBefore = notBefore;
        return this;
    }

    public Instant getNotBefore() {
        return notBefore;
    }

    /**
     * Sets the not-after property.
     * 
     * @param notAfter The not-after value to assign.
     *
     * @return A reference to this API so it can be used fluently.
     */
    public TrustedCertificateAuthority setNotAfter(final Instant notAfter) {
        this.notAfter = notAfter;
        return this;
    }

    public Instant getNotAfter() {
        return notAfter;
    }

    /**
     * Sets the algorithm property of the trusted root certificate.
     * 
     * @param keyAlgorithm  The algorithm name of the public key.
     * @return              a reference to this for fluent use.
     */
    public TrustedCertificateAuthority setKeyAlgorithm(final String keyAlgorithm) {
        this.keyAlgorithm = keyAlgorithm;
        return this;
    }

    public String getKeyAlgorithm() {
        return keyAlgorithm;
    }
}
```

Reply:
The difference between the test case running in `service-base` and the test case running in `core` is the JVM. In `service-base` it is Java 11, in `core` it is Java 8. Run the test in `core` using Java 11, I get the same (expected) result as in `service-base`.
That said, I believe that the deserializer used in `HonoTimestamp` is wrong. It should be `DateTimeFormatter.ISO_OFFSET_DATE_TIME` instead of `DateTimeFormatter.ISO_INSTANT`.
I will test this ...

Reply:
you're really amazing! yes, that was exactly it. thank you 🙏